src/main/assets/js/session-timeout.ts

```diff
@@ -0,0 +1,42 @@
+import { throttle } from 'lodash';
+
+import { TIMED_OUT_URL } from '../../steps/urls';
+
+const eventThrottleTimer = 5 * 60 * 1000; // 5 minutes
+const sessionTimeoutInterval = 20 * 60 * 1000; // 20 minutes
+let timeout;
+
+const saveBeforeSessionTimeout = async () => {
+  const form = document.getElementById('main-form');
+  if (form) {
+    const formData = new FormData(form as HTMLFormElement);
+    formData.append('saveBeforeSessionTimeout', 'true');
+    const url = window.location.pathname;
+    const body = Object.fromEntries(formData);
+    await fetch(url, {
+      method: 'POST',
+      headers: {
+        'Content-type': 'application/json',
+        'csrf-token': body._csrf as string,
+      },
+      body: JSON.stringify(body),
+    });
+  }
+  window.location.href = TIMED_OUT_URL;
+};
+
+const setSaveTimeout = () => {
+  clearTimeout(timeout);
+  timeout = setTimeout(() => {
+    saveBeforeSessionTimeout();
+  }, sessionTimeoutInterval);
+};
+
+const pingUserActive = throttle(() => {
+  fetch('/active').then(() => {
+    setSaveTimeout();
+  });
+}, eventThrottleTimer);
+
+['click', 'touchstart', 'mousemove', 'keypress'].forEach(evt => document.addEventListener(evt, pingUserActive));
+setSaveTimeout();
```

Comment:
Would be nice to get some unit tests on this. I think you should be able to trigger a `keypress`, `click`, `mousemove` etc using Jest and then using Jest mock timers to advance the clock by 20 minutes you should be able to check if `fetch` was called with the right arguments.

Reply:
this is becoming a bit of a pain hence i might merge this once you approve and do this in another pr